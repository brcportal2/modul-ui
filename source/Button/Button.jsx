import React, {Component} from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';

export default class Button extends Component {
	static propTypes = {
		loading: PropTypes.bool,
		disabled: PropTypes.bool,
		className: PropTypes.string,
		icon: PropTypes.string,

		mdStyle: PropTypes.oneOf([
			'second',
			'cancel',
			'dark',
			'white',
			'clean',
			'grey',
			'light',
			'light_primary',
			'light-white',
			'_gd-orange',
		]),

		mdSize: PropTypes.oneOf([
			'xxsmall',
			'xsmall',
			'small',
			'middle',
			'full',
			'wide',
		]),

		type: PropTypes.oneOf([
			'button',
			'reset',
			'submit'
		]),
	};

	static defaultProps = {
		type: 'button',
		loading: false,
		disabled: false
	};

	render() {
		const {
			className,
			loading,
			disabled,
			children,
			mdStyle,
			mdSize,
			icon,
			...props
		} = this.props;

		return (
			<button
				{...props}
				disabled={loading || disabled}

				className={classnames('button', className, mdStyle, mdSize, {
					loader: loading,
					[`icon-${icon}`]: Boolean(icon)
				})}
			>{children}</button>
		);
	}
}