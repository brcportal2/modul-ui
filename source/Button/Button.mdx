---
name: Button
menu: Components
---

import { Playground, PropsTable } from 'docz'
import Button from './Button.jsx'

# Button

Кнопки делают общие действия более очевидными и помогают пользователям более легко выполнять их. Кнопки используют текст и иногда значки, чтобы сообщить о действии, которое произойдет, когда пользователь коснется их.

## Properties

<PropsTable of={Button} />

## Basic usage

<Playground>
  <Button>Click me</Button>
</Playground>

## With different types

<Playground>
  <Button type="button">Click me</Button>
  <Button type="reset">Click me</Button>
  <Button type="submit">Click me</Button>
</Playground>

## With different sizes

<Playground>
  <Button mdSize="xxsmall">Click me</Button>
  <Button mdSize="xsmall">Click me</Button>
  <Button mdSize="small">Click me</Button>
  <Button mdSize="middle">Click me</Button>
  <Button mdSize="full">Click me</Button>
  <Button mdSize="wide">Click me</Button>
</Playground>

## With different styles

<Playground>
  <Button mdStyle="second">Click me</Button>
  <Button mdStyle="cancel">Click me</Button>
  <Button mdStyle="dark">Click me</Button>
  <Button mdStyle="white">Click me</Button>
  <Button mdStyle="clean">Click me</Button>
  <Button mdStyle="grey">Click me</Button>
  <Button mdStyle="light">Click me</Button>
  <Button mdStyle="light_primary">Click me</Button>
  <Button mdStyle="light-white">Click me</Button>
  <Button mdStyle="_gd-orange">Click me</Button>
</Playground>

## Loading

<Button mdSize="middle" loading>Click me</Button>

## Disabled
<Button mdSize="middle" disabled>Click me</Button>