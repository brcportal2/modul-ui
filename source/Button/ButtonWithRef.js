import React from 'react';
import Button from './Button';

const ButtonWithRef = React.forwardRef((props, ref) => <Button {...props} innerRef={ref} />);

ButtonWithRef.displayName = 'ButtonWithRef';
ButtonWithRef.propTypes = Button.propTypes;

export default ButtonWithRef;