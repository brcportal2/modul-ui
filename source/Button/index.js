import Button from './Button';
import ButtonWithRef from './ButtonWithRef';

export default Button;
export {
	Button,
	ButtonWithRef
};