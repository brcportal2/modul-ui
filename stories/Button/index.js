import React from 'react';
import {storiesOf} from '@storybook/react';
import {Button} from '../../source/Button';
import {boolean, select, text} from '@storybook/addon-knobs';
import StoryLayout from '../StoryLayout';

export default storiesOf('Button', module)
	.add(
		'default',

		() =>
			<StoryLayout>
				<Button
					loading={boolean('loading', false)}
					disabled={boolean('disabled', false)}
					wide={boolean('wide', false)}

					mdStyle={select(
						'mdStyle',

						[
							undefined,
							'second',
							'cancel',
							'dark',
							'white',
							'clean',
							'grey',
							'light',
							'light_primary',
							'light-white',
							'_gd-orange',
						],

						undefined
					)}

					mdSize={select(
						'mdSize',

						[
							undefined,
							'xxsmall',
							'xsmall',
							'small',
							'middle',
							'full',
						],

						undefined
					)}

					icon={text('icon', undefined)}
				>
					Hello Button
				</Button>
			</StoryLayout>,

		{info: {text: `
		import {Button} from 'modul-ui';
		`}}
	);