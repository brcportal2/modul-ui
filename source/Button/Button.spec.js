import React from 'react';
import renderer from 'react-test-renderer';

import Button from './Button';


describe('Button', () => {
    test('Default props', () => {
        const component = renderer.create(
            <Button>Test</Button>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Loading', () => {
        const component = renderer.create(
            <Button
                loading
            >Test</Button>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Loading && className', () => {
        const component = renderer.create(
            <Button
                className='test_class'
                loading
            >Test</Button>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Disabled', () => {
        const component = renderer.create(
            <Button
                disabled
            >Test</Button>
        );
        expect(component.toJSON()).toMatchSnapshot();
    });

    test('Type', () => {
        const component_button = renderer.create(
            <Button
                type='button'
            >Test</Button>
        );
        expect(component_button.toJSON()).toMatchSnapshot();

        const component_reset = renderer.create(
            <Button
                type='reset'
            >Test</Button>
        );
        expect(component_reset.toJSON()).toMatchSnapshot();

        const component_submit = renderer.create(
            <Button
                type='submit'
            >Test</Button>
        );
        expect(component_submit.toJSON()).toMatchSnapshot();
    });
});