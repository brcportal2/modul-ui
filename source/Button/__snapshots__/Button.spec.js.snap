// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Default props 1`] = `
<button
  className="button"
  disabled={false}
  type="button"
>
  Test
</button>
`;

exports[`Button Disabled 1`] = `
<button
  className="button"
  disabled={true}
  type="button"
>
  Test
</button>
`;

exports[`Button Loading && className 1`] = `
<button
  className="button test_class loader"
  disabled={true}
  type="button"
>
  Test
</button>
`;

exports[`Button Loading 1`] = `
<button
  className="button loader"
  disabled={true}
  type="button"
>
  Test
</button>
`;

exports[`Button Type 1`] = `
<button
  className="button"
  disabled={false}
  type="button"
>
  Test
</button>
`;

exports[`Button Type 2`] = `
<button
  className="button"
  disabled={false}
  type="reset"
>
  Test
</button>
`;

exports[`Button Type 3`] = `
<button
  className="button"
  disabled={false}
  type="submit"
>
  Test
</button>
`;
